File: 189. Rotate Array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k %= n;
        if (k == 0) return;
        int gcd = gcd(k,n);
        if (gcd == 1) {
            int temp = nums[0];
            for (int j = k; j != 0; j = (j+k)%n) {
                int t = nums[j];
                nums[j] = temp;
                temp = t;
            }
            nums[0] = temp;
        } else {
            for (int i = 0; i < gcd; i++) {
                int temp = nums[i];
                for (int j = i+k; j != i; j = (j+k)%n) {
                    int t = nums[j];
                    nums[j] = temp;
                    temp = t;
                }
                nums[i] = temp;
            }
        }
    }
    private int gcd(int a, int b) {
        if (b % a == 0) return a;
        return gcd(b%a, a);
    }
}

File: 189. Rotate Array.go
func rotate(nums []int, k int)  {
    for k > len(nums) {
        k = k - len(nums)
    }
    gcd := gcd(len(nums), k)
    start := 0
    for start < gcd {
        helper(&nums, k, start)
        start++
    }
}

func helper(nums *[]int, k, i int) {
    start := i
    n := len(*nums)
    i = (i-k+n)%n
    temp := (*nums)[i]
    for i != start {
        (*nums)[i] = (*nums)[(i-k+n)%n]
        i = (i-k+n)%n
    }
    (*nums)[start] = temp
}

func gcd(a, b int) int {
    if a > b {
        return gcd(b, a)
    }
    if a == 0 {
        return b
    }
    return gcd(b % a, a)
}
